### PR TITLE
Fully type annotate `pybreaker.py`

### PR DESCRIPTION
**Pull request by Kludex**
Fully type annotate `pybreaker` and enforce it with `mypy` 😎 👍 

I've added comments on things that are not trivial or fully known stuff. Hope it helps, but if you miss something, I don't mind explaining :)

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -7,3 +7,10 @@ repos:
     rev: 5.10.1
     hooks:
       - id: isort
+  - repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v0.971
+    hooks:
+      - id: mypy
+        pass_filenames: false
```
Comment (Kludex) [inline .pre-commit-config.yaml:14]: We don't want to evaluate `src/tests.py`. Only the files on `files` section on the `pyproject.toml`.

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+[tool.mypy]
+disallow_untyped_defs = true
+ignore_missing_imports = true
+no_implicit_optional = true
+show_error_codes = true
+files = ["src/pybreaker.py"]
+
+[tool.isort]
+profile = "black"
```
Comment (Kludex) [inline pyproject.toml:9]: I forgot this when I added `black` and `isort`.

If this is not set, they may mismatch.

File: src/pybreaker.py [hunk 3]
```diff
@@ -5,15 +5,30 @@
 For more information on this and other patterns and best practices, buy the
 book at https://pragprog.com/titles/mnee2/release-it-second-edition/
 """
+from __future__ import annotations
```
Comment (Kludex) [inline src/pybreaker.py:8]: This is a future flag, you can read more about it on:
- https://lukasz.langa.pl/61df599c-d9d8-4938-868b-36b67fdb4448/
- https://docs.python.org/3/library/__future__.html

But as a summary... It makes the interpreter to not see the type annotations at runtime. Meaning that we gain performance, and we can use syntax of further Python versions (as the IDE is prepared for it).

File: src/pybreaker.py [hunk 4]
```diff
@@ -44,8 +61,13 @@
 STATE_CLOSED = "closed"
 STATE_HALF_OPEN = "half-open"
 
+T = TypeVar("T")
+ExceptionType = TypeVar("ExceptionType", bound=BaseException)
+CBListenerType = TypeVar("CBListenerType", bound="CircuitBreakerListener")
```
Comment (Kludex) [inline src/pybreaker.py:66]: `bound` is used to say: "all subclasses are valid". Without this, it's not possible to type annotate fields that accept objects from subclasses.

File: src/pybreaker.py [hunk 5]
```diff
@@ -44,8 +61,13 @@
...
+CBStateType = Union["CircuitClosedState", "CircuitHalfOpenState", "CircuitOpenState"]
+
 
-class CircuitBreaker(object):
```
Comment (Kludex) [inline src/pybreaker.py:48]: `(object)` is not needed anymore in Python 3.x.

File: src/pybreaker.py [hunk 6]
```diff
@@ -149,11 +176,11 @@ def state(self):
             # changed elsewhere (e.g. another process instance). We still send
             # out a notification, informing others that this particular circuit
             # breaker instance noticed the changed circuit.
-            self.state = self.current_state
+            self.state = self.current_state  # type: ignore[assignment]
```
Comment (Kludex) [inline src/pybreaker.py:179]: I had to ignore some assignments, as mypy has a tracked bug around it: https://github.com/python/mypy/issues/3004 🙏

File: src/pybreaker.py [hunk 7]
```diff
@@ -221,15 +248,15 @@ def is_system_error(self, exception):
...
+    def call(self, func: Callable[..., T], *args: Any, **kwargs: Any) -> T:
         """
         Calls `func` with the given `args` and `kwargs` according to the rules
         implemented by the current state of this circuit breaker.
         """
         with self._lock:
             return self.state.call(func, *args, **kwargs)
 
-    def call_async(self, func, *args, **kwargs):
+    def call_async(self, func, *args, **kwargs):  # type: ignore[no-untyped-def]
```
Comment (Kludex) [inline src/pybreaker.py:259]: I didn't add any type annotation to code related to `tornado`. 👀

File: src/pybreaker.py [hunk 8]
```diff
@@ -701,6 +723,16 @@ def name(self) -> str:
...
 
+    @overload
+    def _handle_error(
+        self, exc: BaseException, reraise: Literal[True] = ...
+    ) -> NoReturn:
+        ...
+
+    @overload
+    def _handle_error(self, exc: BaseException, reraise: Literal[False] = ...) -> None:
+        ...
```
Comment (Kludex) [inline src/pybreaker.py:734]: If `reraise` is `True`, then we never return a value i.e. always exception.

File: src/pybreaker.py [hunk 9]
```diff
@@ -935,5 +979,3 @@ class CircuitBreakerError(Exception):
     When calls to a service fails because the circuit is open, this error is
     raised to allow the caller to handle this type of exception differently.
     """
-
-    pass
```
Comment (Kludex) [inline src/pybreaker.py:939]: `pass` are not needed when docstring is present.

File: setup.py [hunk 10]
```diff
@@ -22,13 +22,9 @@
     package_dir={"": "src"},
     py_modules=["pybreaker"],
     include_package_data=True,
+    install_requires=["typing-extensions>=3.10.0; python < '3.8'"],
```
Comment (Kludex) [inline setup.py:25]: [`Literal`](https://docs.python.org/3/library/typing.html#typing.Literal) was introduced in Python 3.8. `typing_extensions` is the official backport package for typing classes. 

Python 3.7 is almost at EOL (December 2022).

**Comment by Kludex** [comment]
Also, would you mind publishing a release after this? 🙏

**Comment by danielfm** [comment]
Would these changes somehow break code that don't use type annotations? Since I'm not familiar with those, I thought I'd better ask. 😅

**Comment by Kludex** [comment]
> Would these changes somehow break client code (code that uses pybreaker) that don't use type annotations? Since I'm not familiar with those, I thought I'd better ask. 😅

No, it doesn't break client code.

Also, the only thing that I changed that is not type related on this PR is the "calling parent constructor" way, which is redundant on Python 3.x. In other words, it's not even a breaking change for `pybreaker` iteself.

File: src/pybreaker.py (hunk 4 above)
Comment (danielfm) [reply src/pybreaker.py:70]: Makes sense, thanks for the explanation.

**Comment by danielfm** [comment]
Not sure I quite like the way Python evolved as a language, these type annotations make the code a lot noiser for my taste, but well, since this is where Python is going, we shall follow. 😄 

Thanks for your time, your contributions were greatly appreciated!

**Comment by Kludex** [comment]
I cannot work without autocompletion, and this PR enables that. 🙏

Also, the type checker helps us to find issues along the way (in this case there were none 🤷‍♂️ ).

Thanks for accepting those PRs! :)

**Pull request merged**